requests/models.py

```diff
@@ -470,18 +469,12 @@ def prepare_body(self, data, files, json=None):
         self.body = body
 
     def prepare_content_length(self, body):
-        if hasattr(body, 'seek') and hasattr(body, 'tell'):
-            curr_pos = body.tell()
-            body.seek(0, 2)
-            end_pos = body.tell()
-            self.headers['Content-Length'] = builtin_str(max(0, end_pos - curr_pos))
-            body.seek(curr_pos, 0)
-        elif body is not None:
-            l = super_len(body)
-            if l:
-                self.headers['Content-Length'] = builtin_str(l)
-        elif (self.method not in ('GET', 'HEAD')) and (self.headers.get('Content-Length') is None):
-            self.headers['Content-Length'] = '0'
+        if self.headers.get('Content-Length') is None:
+            if body is not None:
+                length = super_len(body)
+                self.headers['Content-Length'] = builtin_str(length)
+            elif self.method not in ('GET', 'HEAD'):
+                self.headers['Content-Length'] = '0'
```

Comment:
Sorry, this probably should have been split into two PRs but I'll try to define the problems more clearly. 

This is the "solution" that was closest to the original but I would say not a complete or ideal one. I was looking for some more input on what direction you felt was best to proceed in. We definitely need to move `if self.headers.get('Content-Length') is None` outside of this method, or change it to `if self.headers.get('Content-Length') is None or overwrite` for this to work properly.

Currently, we overwrite 'Content-Length' in ALL cases where a body is attached. This will subtly override a 'Content-Length' header passed to a request. @kennethreitz stated in #2329 that he'd prefer users be able to do this and I feel like having to go back and change 'Content-Length' after already passing your custom value is awkward.

Some examples:

```python
>>> import requests
>>> r = requests.get('http://httpbin.org/get', headers={'Content-Length': '5'}, data="thisisatest")
>>> r.request.headers
{'Connection': 'keep-alive', 'Content-Length': '11', 'Accept-Encoding': 'gzip, deflate', 'Accept': '*/*', 'User-Agent': 'python-requests/2.11.1'}
```

```python
>>> import requests
>>> r = requests.Request('GET', 'http://httpbin.org/get', headers={'Content-Length': '5'}, data="thisisatest")
>>> r.headers
{'Content-Length': '5'}
>>> pr = r.prepare()
>>> pr.headers
{'Content-Length': '11'}
```

Having to do `pr.headers['Content-Length'] = '5'` after I've already passed in the custom headers I want is the issue I'm trying to address. If this is a feature we're going to support, does it need to subtly "fail" like this.